README.md
[![Build Status](https://secure.travis-ci.org/customink-webops/stoplight.png?branch=master)](http://travis-ci.org/customink-webops/stoplight)

Description
===========
Stoplight is a build monitoring tool that is largely based off greenscreen, but is much improved and expandable. To quickly name a few, Stoplight has:

* built-in support for Jenkins
* built-in support for Travis-CI (http://travis-ci.org)
* custom provider support
* community contributions
* full test suite
* resuable DSL

Stoplight is designed to be displayed on large television screens or monitors. It automatically resizes to fill the maximum real estate the screen can offer.

This cookbook configures a node to run one or more Stoplight applications.  The cookbook uses the [Stoplight project](https://github.com/customink/stoplight).

Requirements
============

## Cookbooks:

Opscode Cookbooks (http://community.opscode.com/cookbooks)

* git


Attributes
==========

* `default['stoplight']['install_dir']` - Location where the Stoplight application will run
* `default['stoplight']['servers']` - An array of Stoplight configuration instances
* `default['stoplight']['servers']['name']` - The name of this Stoplight
* `default['stoplight']['servers']['port']` - The port used by this Stoplight
* `default['stoplight']['servers']['servers']` - An array of servers (providers) that this Stoplight should watch
* `default['stoplight']['servers']['servers']['url']` - The URL for this build server
* `default['stoplight']['servers']['servers']['username']` - The login for this server (optional)
* `default['stoplight']['servers']['servers']['password']` - The password for this server (optional)
* `default['stoplight']['servers']['servers']['projects']` - Array of jobs to poll for.  Leave empty to watch all projects
* `default['stoplight']['servers']['servers']['ignored_projects']` - Array of projects to ignore.  Leave empty to watch all projects


License and Authors
===================

Author:: Seth Vargo <svargo@customink.com>

Copyright:: 2012, CustomInk, LLC.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.


CHANGELOG.md
CHANGELOG
=========
v1.0.0
------
- Add test-kitchen and upgrade dependencies

v0.1.4
------
- Restart passenger on deploy

v0.1.3
------
- Use apache and passenger for deployment

v0.1.2
------
- Actually kill server process if it exists
- move config template to reflect actual directory structure

v0.1.1
------
- Fixed typo
- Added chefignore

v0.1.0
------
- Initial Release


recipes/apache2.rb
#
# Cookbook Name:: stoplight
# Recipe:: apache2
#
# Copyright 2012, CustomInk, LLC
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#

include_recipe 'apache2'
include_recipe 'passenger_apache2'

apache_site '000-default' do
  enable false
end

node['stoplight']['servers'].each do |server|
  file = "#{node['apache']['dir']}/sites-available/#{server['name']}"
  template file do
    source 'apache2/site.erb'
    mode 0644
    variables(
      :install_dir => node['stoplight']['install_dir'],
      :server => server,
      :apache => server['apache']
    )
  end

  apache_site server['name'] do
    enable true
  end
end

notifies :reload, 'service[apache2]'


recipes/default.rb
#
# Cookbook Name:: stoplight
# Recipe:: default
#
# Copyright 2012-2013, Seth Vargo (sethvargo@gmail.com)
# Copyright 2012-2013, CustomInk, LLC
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#

include_recipe 'git'
include_recipe 'passenger_apache2'
include_recipe 'passenger_apache2::mod_rails'

gem_package 'bundler'

apache_site '000-default' do
  enable false
end

node['stoplight']['servers'].each do |server|
  server_dir = [ node['stoplight']['install_dir'], server['name'] ].join('/')

  directory server_dir do
    recursive true
    owner node['apache']['user']
    group node['apache']['group']
  end

  git server_dir do
    repository node['stoplight']['repo']
    reference node['stoplight']['revision']
    enable_submodules true
    user node['apache']['user']
    group node['apache']['group']
    action :sync
  end

  execute 'Bundle install application dependencies' do
    cwd server_dir
    command 'bundle install'
  end

  template "#{server_dir}/config/servers.yml" do
    source 'config/servers.yml.erb'
    owner node['apache']['user']
    group node['apache']['group']
    variables(
      :servers => server['servers']
    )
  end

  template "#{node['apache']['dir']}/sites-available/#{server['name']}" do
    source 'apache2/site.erb'
    mode 0644
    variables(
      :server_dir => server_dir,
      :server => server
    )
    notifies :reload, 'service[apache2]'
  end

  apache_site server['name'] do
    enable true
  end

  execute 'Restart passenger' do
    cwd server_dir
    command 'touch tmp/restart.txt'
    action :nothing
  end
end
